pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+		AddFunc:    appMgr.addApp,
+		UpdateFunc: appMgr.updateApp,
+		DeleteFunc: appMgr.deleteApp,
+	})
+	informerFactory.Start(wait.NeverStop)
+
+	utilruntime.Must(appv1.AddToScheme(appscheme.Scheme))
+	return &Controller{
+		kubeClient:       kubeClient,
+		extensionsClient: apiextensionsClient,
+		appClient:        appClient,
+		informer:         informer.Informer(),
+		lister:           informer.Lister(),
+	}
+}
+
+func (c *Controller) Run() {
+	//TODO: handle the case when the crd is not registered: register it programatically
+	defer utilruntime.HandleCrash()
+
+	log.Logger.Info("Waiting cache to be synced.")
+	// Handle timeout for syncing.
+	timeout := time.NewTimer(time.Second * 30)
+	timeoutCh := make(chan struct{})
+	go func() {
+		<-timeout.C
+		timeoutCh <- struct{}{}
+	}()
+	if ok := cache.WaitForCacheSync(timeoutCh, c.informer.HasSynced); !ok {
+		log.Logger.Error("Timeout expired during waiting for caches to sync. Skip starting the application controller")
+	} else {
+		log.Logger.Info("Starting application controller.")
+	}
+}
+
+/*
+Update the application in the scheduler
+*/
+func (appMgr *AppManager) updateApp(oldObj interface{}, newObj interface{}) {
+	newApp, ok := newObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("newObj is not an Application")
+	}
+	oldApp, ok := oldObj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("oldObj is not an Application")
+	}
+	// No need to update if ResourceVersion is not changed
+	if newApp.ResourceVersion == oldApp.ResourceVersion {
+		log.Logger.Info("No need to update because app object is not modified")
+		return
+	}
+
+	// handle spec change
+	if !equality.Semantic.DeepEqual(oldApp.Spec, newApp.Spec) {
+		//TODO: handle update, discuss what should be updated since only the queue is the common property yet
+	}
+}
+
+/*
+Remove the application from the scheduler as well
+*/
+func (appMgr *AppManager) deleteApp(obj interface{}) {
+	app, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	appID := app.Name
+	err := appMgr.amProtocol.RemoveApplication(appID)
+	if err != nil {
+		log.Logger.Error("Application removal failed",
+			zap.String("appID", appID),
+			zap.Error(err))
+	}
+	log.Logger.Debug("App CRD deleted",
+		zap.String("Name", appID))
+	//TODO: delete related pods and deployments
+}
+
+/*
+Add application to scheduler
+*/
+func (appMgr *AppManager) addApp(obj interface{}) {
+	appCRD, ok := obj.(*appv1.Application)
+	if !ok {
+		log.Logger.Error("obj is not an Application")
+	}
+	if appMeta, ok := appMgr.getAppMetadata(appCRD); ok {
+		app := appMgr.amProtocol.GetApplication(appMeta.ApplicationID)
+		if app == nil {
+			appMgr.amProtocol.AddApplication(&interfaces.AddApplicationRequest{
+				Metadata: appMeta,
+				//TODO: debug recovery
+				//TODO: try to define the owner reference for the submitted pods to this CRD and try of we delete the
+				//CRD the pods will be deleted as well
```

Comment:
We need a separate JIRA to track how to implement the recovery, pls file one

Reply:
Filed YUNIKORN-347